File: concept/thing/impl/RelationImpl.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package grakn.client.concept.thing.impl;

import grakn.client.GraknClient;
import grakn.client.concept.thing.Relation;
import grakn.client.concept.thing.Thing;
import grakn.client.concept.type.RoleType;
import grakn.client.concept.type.impl.RelationTypeImpl;
import grakn.client.concept.type.impl.RoleTypeImpl;
import grakn.client.concept.type.impl.TypeImpl;
import grakn.common.collection.Bytes;
import grakn.protocol.ConceptProto;
import grakn.protocol.TransactionProto;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;

import static grakn.client.concept.proto.ConceptProtoBuilder.iid;
import static grakn.client.concept.proto.ConceptProtoBuilder.thing;
import static grakn.client.concept.proto.ConceptProtoBuilder.type;
import static grakn.client.concept.proto.ConceptProtoBuilder.types;

public class RelationImpl extends ThingImpl implements Relation {

    RelationImpl(String iid) {
        super(iid);
    }

    public static RelationImpl of(ConceptProto.Thing protoThing) {
        return new RelationImpl(Bytes.bytesToHexString(protoThing.getIid().toByteArray()));
    }

    @Override
    public RelationImpl.Remote asRemote(GraknClient.Transaction transaction) {
        return new RelationImpl.Remote(transaction, getIID());
    }

    @Override
    public final RelationImpl asRelation() {
        return this;
    }

    public static class Remote extends ThingImpl.Remote implements Relation.Remote {

        public Remote(GraknClient.Transaction transaction, String iid) {
            super(transaction, iid);
        }

        @Override
        public RelationImpl.Remote asRemote(GraknClient.Transaction transaction) {
            return new RelationImpl.Remote(transaction, getIID());
        }

        @Override
        public RelationTypeImpl getType() {
            return super.getType().asRelationType();
        }

        @Override
        public Map<RoleTypeImpl, List<ThingImpl>> getPlayersByRoleType() {
            ConceptProto.Thing.Req.Builder method = ConceptProto.Thing.Req.newBuilder()
                    .setRelationGetPlayersByRoleTypeReq(ConceptProto.Relation.GetPlayersByRoleType.Req.getDefaultInstance());

            TransactionProto.Transaction.Req.Builder request = TransactionProto.Transaction.Req.newBuilder()
                    .setThingReq(method.setIid(iid(getIID())));
            Stream<ConceptProto.Relation.GetPlayersByRoleType.RoleTypeWithPlayer> stream = transactionRPC.stream(
                    request, res -> res.getThingRes().getRelationGetPlayersByRoleTypeRes().getRoleTypesWithPlayersList().stream());

            Map<RoleTypeImpl, List<ThingImpl>> rolePlayerMap = new HashMap<>();
            stream.forEach(rolePlayer -> {
                RoleTypeImpl role = TypeImpl.of(rolePlayer.getRoleType()).asRoleType();
                ThingImpl player = ThingImpl.of(rolePlayer.getPlayer());
                if (rolePlayerMap.containsKey(role)) {
                    rolePlayerMap.get(role).add(player);
                } else {
                    rolePlayerMap.put(role, new ArrayList<>(Collections.singletonList(player)));
                }
            });
            return rolePlayerMap;
        }

        @Override
        public Stream<ThingImpl> getPlayers(RoleType... roleTypes) {
            return thingStream(
                    ConceptProto.Thing.Req.newBuilder().setRelationGetPlayersReq(
                            ConceptProto.Relation.GetPlayers.Req.newBuilder().addAllRoleTypes(types(Arrays.asList(roleTypes)))),
                    res -> res.getRelationGetPlayersRes().getThingsList());
        }

        @Override
        public void addPlayer(RoleType roleType, Thing player) {
            execute(ConceptProto.Thing.Req.newBuilder().setRelationAddPlayerReq(
                    ConceptProto.Relation.AddPlayer.Req.newBuilder().setRoleType(type(roleType)).setPlayer(thing(player))));
        }

        @Override
        public void removePlayer(RoleType roleType, Thing player) {
            execute(ConceptProto.Thing.Req.newBuilder().setRelationRemovePlayerReq(
                    ConceptProto.Relation.RemovePlayer.Req.newBuilder().setRoleType(type(roleType)).setPlayer(thing(player))));
        }

        @Override
        public final RelationImpl.Remote asRelation() {
            return this;
        }
    }
}

File: concept/thing/impl/RelationImpl.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {
    ThingImpl,
    RemoteThingImpl,
    Relation,
    RemoteRelation,
    Thing,
    RelationTypeImpl,
    RoleType,
    GraknClient,
    Stream,
    TransactionRPC,
    RoleTypeImpl, ConceptProtoBuilder, ThingTypeImpl, TypeImpl, Bytes,
} from "../../../dependencies_internal";
import Transaction = GraknClient.Transaction;
import ConceptProto from "grakn-protocol/protobuf/concept_pb";
import TransactionProto from "grakn-protocol/protobuf/transaction_pb";

export class RelationImpl extends ThingImpl implements Relation {
    private readonly _type: RelationTypeImpl;

    protected constructor(iid: string, type: RelationTypeImpl) {
        super(iid);
        this._type = type;
    }

    static of(protoThing: ConceptProto.Thing): RelationImpl {
        return new RelationImpl(Bytes.bytesToHexString(protoThing.getIid_asU8()), RelationTypeImpl.of(protoThing.getType()));
    }

    asRemote(transaction: Transaction): RemoteRelationImpl {
        return new RemoteRelationImpl(transaction, this.getIID(), this._type);
    }

    getType(): RelationTypeImpl {
        return this._type;
    }

    isRelation(): boolean {
        return true;
    }
}

export class RemoteRelationImpl extends RemoteThingImpl implements RemoteRelation {
    private readonly _type: RelationTypeImpl;

    constructor(transaction: Transaction, iid: string, type: RelationTypeImpl) {
        super(transaction, iid);
        this._type = type;
    }

    asRemote(transaction: Transaction): RemoteRelationImpl {
        return new RemoteRelationImpl(transaction, this.getIID(), this._type);
    }

    getType(): RelationTypeImpl {
        return this._type;
    }

    async getPlayersByRoleType(): Promise<Map<RoleType, Thing[]>> {
        const method = new ConceptProto.Thing.Req()
            .setRelationGetPlayersByRoleTypeReq(new ConceptProto.Relation.GetPlayersByRoleType.Req())
            .setIid(Bytes.hexStringToBytes(this.getIID()));
        const request = new TransactionProto.Transaction.Req().setThingReq(method);
        const stream = (this.transaction as TransactionRPC).stream(request, res => res.getThingRes().getRelationGetPlayersByRoleTypeRes().getRoleTypesWithPlayersList());
        const rolePlayerMap = new Map<RoleTypeImpl, ThingImpl[]>();
        for await (const rolePlayer of stream) {
            const role = TypeImpl.of(rolePlayer.getRoleType()) as RoleTypeImpl;
            const player = ThingImpl.of(rolePlayer.getPlayer());
            let addedToExistingEntry = false;
            for (const roleKey of rolePlayerMap.keys()) {
                if (roleKey.getScopedLabel() === role.getScopedLabel()) {
                    rolePlayerMap.get(roleKey).push(player);
                    addedToExistingEntry = true;
                    break;
                }
            }
            if (!addedToExistingEntry) rolePlayerMap.set(role, [player]);
        }
        return rolePlayerMap;
    }

    getPlayers(roleTypes: RoleType[] = []): Stream<ThingImpl> {
        const method = new ConceptProto.Thing.Req().setRelationGetPlayersReq(
            new ConceptProto.Relation.GetPlayers.Req().setRoleTypesList(roleTypes.map(roleType => ConceptProtoBuilder.type(roleType))));
        return this.thingStream(method, res => res.getRelationGetPlayersRes().getThingsList()) as Stream<ThingImpl>;
    }

    async addPlayer(roleType: RoleType, player: Thing): Promise<void> {
        await this.execute(new ConceptProto.Thing.Req().setRelationAddPlayerReq(
            new ConceptProto.Relation.AddPlayer.Req()
                .setPlayer(ConceptProtoBuilder.thing(player))
                .setRoleType(ConceptProtoBuilder.type(roleType))
        ));
    }

    async removePlayer(roleType: RoleType, player: Thing): Promise<void> {
        await this.execute(new ConceptProto.Thing.Req().setRelationRemovePlayerReq(
            new ConceptProto.Relation.RemovePlayer.Req()
                .setPlayer(ConceptProtoBuilder.thing(player))
                .setRoleType(ConceptProtoBuilder.type(roleType))
        ));
    }

    isRelation(): boolean {
        return true;
    }
}
